libraries/AC_Avoidance/AP_OAVisGraph.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+
+/*
+ * Visibility graph used by Dijkstra's algorithm for path planning around fence, stay-out zones and moving obstacles
+ */
+class AP_OAVisGraph {
+public:
+    AP_OAVisGraph();
+    AP_OAVisGraph(uint8_t size);
+
+    /* Do not allow copies */
+    AP_OAVisGraph(const AP_OAVisGraph &other) = delete;
+    AP_OAVisGraph &operator=(const AP_OAVisGraph&) = delete;
+
+    // types of items held in graph
+    enum OAType : uint8_t {
+        OATYPE_SOURCE = 0,
+        OATYPE_DESTINATION,
+        OATYPE_FENCE_POINT
+    };
+
+    // support up to 255 items of each type
+    typedef uint8_t oaid_num;
+
+    // id for uniquely identifying objects held in visibility graphs and paths
+    class OAItemID {
+    public:
+        OAType id_type;
+        oaid_num id_num;
+        bool operator ==(const OAItemID &i) const { return ((id_type == i.id_type) && (id_num == i.id_num)); }
+    };
+
+    struct VisGraphItem {
+        OAItemID id1;       // first item's id
+        OAItemID id2;       // second item's id
+        float distance_cm;  // distance between the items
+    };
+
+    // initialise array to given size
+    bool init(uint8_t size);
+
+    // clear all elements from graph
+    void clear() { _num_items = 0; }
+
+    // get number of items in visibility graph table
+    uint8_t num_items() const { return _num_items; }
+
+    // add item to visiblity graph, returns true on success, false if graph is full
+    bool add_item(const OAItemID &id1, const OAItemID &id2, float distance_cm);
+
+    // get item by index number
+    const VisGraphItem& get_item(uint8_t index_num) const { return _items[index_num]; }
+
+    // allow accessing graph as an array, 0 indexed
+    VisGraphItem& operator[](uint8_t i) { return _items[i]; }
+    const VisGraphItem& operator[](uint8_t i) const { return _items[i]; }
```

Comment:
If you failed to init and allocate the array this will be a nullpointer dereference.

Reply:
the VisGraph::operator[] it's a little dangerous perhaps but at the moment we only use it within loops that use the num_items() function so they will never try to access invalid memory.  I'll add a note to warn that there are no protections.